unifi_mqtt/translator.py

```diff
@@ -31,9 +32,24 @@ def format_target(target_list):
 
 
 def serialize_network(event, payload):
+    # the character set used for cleaning up MQTT topic names
+    mqtt_character_whitelist = set().union(string.ascii_letters, string.digits, string.punctuation)
```

Comment:
i wonder if we should just regexp this and apply it to basically everything (e.g. there's the format_name clause as well that could be replaced)

something like `^[_\s]+` => `-`, `^[a-z0-9\.]+' => `''`

Reply:
`:` is also valid, for example. According to [MQTT specs](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718109) _"Topic Names and Topic Filters are UTF-8 encoded strings, are case sensitive, can include the space character and must not include the null character"_. That's why I used this specific combination of `string.*` character sets. It could make sense to implement this as a more globally available function. A version based on `re.sub()` would work, too. Do you prefer that over the lambda functions?

Reply:
What about

```python3.8
def format_name(name):
    name = name.replace(".", "").replace(" ", "-").replace("_", "-").lower()
    return re.sub(rf"[^{string.ascii_letters}{string.digits}:\-\s]+", "-", name).rstrip('-')
```

and

```python3.8
client_name = format_name(payload["hostname"])
```

Reply:
yeah we could do something like that - i would probably use a-z0-9 tho for regexp if we're gonna do regex (since its a guarantee and everyone would know what it is). Only reason I suggested it as I imagine there will be performance implications given the throughput of data these systems could possible see, and I believe the regexp parser will be faster. Possible thats not true though.

Reply:
As far as I know, `re` uses some C code for matching, so that should in theory be significantly faster than anything else. In terms of pattern and speed optimization, I can offer you to
- remove the `string` import while at the same time making the pattern more human-readable
- compile the pattern outside the function since we're re-using it
- converting the `.replace()` chain to a single use of `.translate()` which relies on a translation table defined outside the function
- include `A-Z` alongside `a-z0-9` since MQTT topics are case sensitive

The code would then look like this:
```python3.8
import re

mqtt_translation_table = str.maketrans({".": "", " ": "-", "_": "-"})
mqtt_regex_object = re.compile(r"[^a-zA-Z0-9\:\-\s]+")

def format_name(name):
    name = name.translate(mqtt_translation_table).lower()
    return mqtt_regex_object.sub("-", name).rstrip('-')
```

and then this
```python3.8
print(format_name('test name_With//(&/%/(sy:::mbo...ls!everyWhere-!'))
```
would result in
```string
test-name-with-sy:::mbols-everywhere
```

Reply:
i think in general either approach sounds fine (as long as we match chars) and we can always tweak later if perf is an issue